draft/XXXX-upload-system/README.md

```diff
@@ -0,0 +1,126 @@
+    tables: string[],
+    graphs: string[],
+  },
+  user: user_id (sub value),
+  _id: id (arangodb, used by clients to actually send chunks)
+}
+```
+
+### Initializing an Upload
+To initialize an upload, new endpoints will be needed, which will replace the existing upload endpoints. To initialize an upload, a request of the following format is made (endpoint for each upload format specified):
+
+```
+POST /api/upload/csv/{workspace}/{table}
+POST /api/upload/d3_json/{workspace}/{graph}
+POST /api/upload/nested_json/{workspace}/{graph}
+POST /api/upload/newick/{workspace}/{graph}
+```
+
+These endpoints replaces the original upload endpoints for each data type we currently support. This returns a JSON response, containing the upload information, in the format referenced above.
+
+A new endpoint for uploading data chunks themselves will have the following format (endpoint for each upload format specified):
+
+```
+PUT /api/upload/csv/{workspace}/{table}
+PUT /api/upload/d3_json/{workspace}/{graph}
+PUT /api/upload/nested_json/{workspace}/{graph}
+PUT /api/upload/newick/{workspace}/{graph}
+
+Body: A chunk of the original data
+
+Query Parameters: {
+  upload: id (returned by the first call),
+}
+```
+
+The `upload` query parameter is part of the information returned by the first call, and is used to denote the upload object to which the sent chunk of data belongs to. Since an upload is tied to a specific user, this endpoint is gated behind user login. Furthermore, a `PUT` request (by a logged in user) to an upload endpoint targeting an upload that does not belong to this user, should return `403 Forbidden`.
+
+
+New endpoints for getting upload information are also required. These would be of the following formats:
+
+```
+GET /api/upload/uploads
+```
+
+This endpoint is tied to the currently authenticated user, and returns all of the uploads which belong to the user. If the need for a global upload listing endpoint arises, this may have to be revised.
+
+
+### Intermediate Processing of Chunks
+
+Each data type uploader (csv, newick, etc.) will be responsible for specific logic regarding how chunked uploads are handled and stored in the "hidden" arangodb collection. Given a chunk of data, the uploader needs to be able to process/keep track the following information:
+
+* How many rows/documents are fully contained in that chunk
+* How to convert/transform a row/document from its original format, into the format required for arangodb storage
+* The current position or "context" in the full document (in graphs for example, are we currently looking at nodes?, links?, etc.)
+
+More than likely, a chunk will not often start/end exactly on the boundary between rows/documents. In this case, any "incomplete" rows/documents will need to stored in a specific location (perhaps on the upload document) until the data which completes that record arrives. Storing this data in memory is not an option, as compatibility across multiple workers must be ensured.
```

Comment:
This is the diciest part of this RFC in my opinion, and it feels like we just don't have enough understanding of traditional multipart upload schemes.

One "simple" way (air quotes because I am not sure it's this simple) is to enforce these data boundaries on the uploader logic itself. That is, the d3-json uploader would need to scan the data to deliver only completed JSON objects in each chunk. This seems pretty difficult in general, when it just feels like there ought to be a general method for collecting chunks of raw data on the server before executing the upload logic.

We will want to ask @zachmullen or other Girder wizards about this.

Reply:
If you want naive chunked upload, I would look to S3 as a good approximation of the state of the art. First there's an init request to create a unique ID for the upload, which validates the name (and in your case could reserve it atomically). Then chunks (S3 calls them Parts) can be uploaded in parallel and are just byte sequences that can be any size within a lower and upper bound. Once all chunks are uploaded, the client sends a completion document that lists the parts in order, and the server is responsible for "stitching" them together if you will. Basically each Part is a file with a special flag indicating it's a part, and these may be stitched together arbitrarily with a finalization request.

If you're wanting something less naive, I think the client-side chunking to split along data boundaries is a good idea. It's really not that hard to achieve.